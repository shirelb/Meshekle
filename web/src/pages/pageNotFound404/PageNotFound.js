import React from 'react';
import {Grid, Header, Image} from 'semantic-ui-react';

const PageNotFound = () => (
    <Grid stretched textAlign={"center"}>
        <Grid.Column stretched textAlign={"center"}>
            <Image centered src={require('../../images/logo1.png')} size={'large'}/>
            <Header size={"huge"} textAlign={"center"} icon style={{marginTop: "-10px"}}>
                משקל'ה
                <Header.Subheader>יש משק ויש משקל'ה</Header.Subheader>
            </Header>

            <div style={{marginTop: 20, textAlign: "center"}}>
                <Header as={'h1'}> 404 </Header>
                <Header as={'h2'}> Page not found </Header>
            </div>
        </Grid.Column>
    </Grid>
);

export default PageNotFound;