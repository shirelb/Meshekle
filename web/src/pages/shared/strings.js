const loginPageStrings = {
    LOGIN: "התחבר",
    SUBMIT: "שלח",
    WELCOME_TO_LOGIN: "Welcome to the login screen!",
    USER_ID_PLACEHOLDER: "ת.ז.",
    PASSWORD_PLACEHOLDER: "סיסמא",
    APP_NAME: "משקל\'ה",
    WRONG_CREDENTIALS: "אחד הפרטים שהזנת לא נכון",
    NO_SUCH_USER: "המשתמש אינו קיים",
    EMPTY_USER_ID: "הוסף ת.ז. ונסה להתחבר שוב",
    EMPTY_PASSWORD: "הוסף סיסמא ונסה להתחבר שוב",
};

const mainScreenStrings = {
    LOGOUT: "התנתק",
    WELCOME: "ברוכים הבאים ",
};

const drawerMenu = {
    LOGOUT: "התנתק",
    MainScreenName: "בית",
    AppointmentsScreenName: "תורים",
    ChoresScreenName: "תורנויות",
};

export default {loginPageStrings, mainScreenStrings, drawerMenu};