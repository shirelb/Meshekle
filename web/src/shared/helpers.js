import axios from "axios";
import {SERVER_URL} from "./constants";
import store from "store";


const serviceProviderHeaders = {
    'Authorization': 'Bearer ' + store.get('serviceProviderToken')
};

var getUserByUserID = (userId, headers) => {
    return axios.get(`${SERVER_URL}/api/users/userId/${userId}`,
        {headers: headers}
    )
        .then((response) => {
            let user = response.data[0];
            return user;
        })
        .catch((error) => {
            console.log('getUserByUserID ', userId, ' ', error);
            return null;
        });
};

var getAppointmentByAppointmentID = (serviceProviderId, appointmentId, headers) => {
    return axios.get(`${SERVER_URL}/api/serviceProviders/appointments/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                appointmentId: appointmentId,
            }
        }
    )
        .then((response) => {
            let appointment = response.data[0];
            console.log('getAppointmentByAppointmentID ', appointmentId, ' ', appointment);
            return appointment;
        })
        .catch((error) => {
            console.log('getAppointmentByAppointmentID ', appointmentId, ' ', error);
            return null;
        });
};

var getAppointmentRequestByAppointmentRequestID = (serviceProviderId, appointmentRequestId, headers) => {
    return axios.get(`${SERVER_URL}/api/serviceProviders/appointmentRequests/serviceProviderId/${serviceProviderId}`,
        {
            headers: headers,
            params: {
                appointmentRequestId: appointmentRequestId,
            }
        }
    )
        .then((response) => {
            let appointmentRequest = response.data[0];
            console.log('getAppointmentByAppointmentID ', appointmentRequestId, ' ', appointmentRequest);
            return appointmentRequest;
        })
        .catch((error) => {
            console.log('getAppointmentByAppointmentID ', appointmentRequestId, ' ', error);
            return null;
        });
};

var getUsers = () => {
    return axios.get(`${SERVER_URL}/api/users`,
        {headers: serviceProviderHeaders}
    )
        .then((response) => {
            return response.data;
        });
};

var getRolesOfServiceProvider = (serviceProviderId) => {
    return axios.get(`${SERVER_URL}/api/serviceProviders/roles/serviceProviderId/${serviceProviderId}`,
        {headers: serviceProviderHeaders}
    )
        .then((response) => {
            let roles = response.data;
            console.log('roles ', roles);
            return roles;
        })
        .catch((error) => {
            console.log('error ', error);
        });
};

export default {
    getUserByUserID,
    getAppointmentByAppointmentID,
    getUsers,
    getRolesOfServiceProvider,
    getAppointmentRequestByAppointmentRequestID
};
