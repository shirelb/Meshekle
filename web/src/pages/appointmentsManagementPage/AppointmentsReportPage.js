import React from 'react';
import './styles.css';

import moment from 'moment';

import {Button, Grid, Header, Icon, Menu, Table} from 'semantic-ui-react';
import 'semantic-ui-css/semantic.min.css';
import store from 'store';
import {Helmet} from 'react-helmet';
import times from 'lodash.times';
import strings from "../../shared/strings";
import AppointmentInfo from "../../components/appointment/AppointmentInfo";
import AppointmentEdit from "../../components/appointment/AppointmentEdit";
import {Link, Route, Switch} from "react-router-dom";
import AppointmentAdd from "../../components/appointment/AppointmentAdd";
import AppointmentRequestInfo from "../../components/appointmentRequest/AppointmentRequestInfo";
import appointmentsStorage from "../../storage/appointmentsStorage";
import usersStorage from "../../storage/usersStorage";
import {connectToServerSocket, WEB_SOCKET} from "../../shared/constants";
import ServiceProviderEdit from "../../components/serviceProvider/ServiceProviderEdit";

const TOTAL_PER_PAGE = 10;


class AppointmentsReportPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            appointments: [],
            appointmentRequests: [],
            calendarEvents: [],
            page: 0,
            totalPages: 0,
            openPopup: false,
            eventPopup: {},
            highlightTableRow: null,
            appointmentRequestHoovering: {requestId: -1},
        };

        this.incrementPage = this.incrementPage.bind(this);
        this.decrementPage = this.decrementPage.bind(this);
        this.setPage = this.setPage.bind(this);
        this.handleDelete = this.handleDelete.bind(this);
        this.updateAfterMoveOrResizeEvent = this.updateAfterMoveOrResizeEvent.bind(this);

        this.serviceProviderHeaders = '';
    }

    componentDidMount() {
        this.serviceProviderHeaders = {
            'Authorization': 'Bearer ' + store.get('serviceProviderToken')
        };
        this.userId = store.get('userId');
        this.serviceProviderId = store.get('serviceProviderId');

        this.getServiceProviderAppointments();
        this.getServiceProviderAppointmentRequests();

        connectToServerSocket(store.get('serviceProviderId'));

        WEB_SOCKET.on("getServiceProviderAppointmentRequests", this.getServiceProviderAppointmentRequests.bind(this));
    }

    componentWillUnmount() {
        WEB_SOCKET.off("getServiceProviderAppointmentRequests");
    }

    getServiceProviderAppointmentRequests() {
        this.setState({
            appointmentRequests: [],
        });

        appointmentsStorage.getServiceProviderAppointmentRequests(this.serviceProviderId, this.serviceProviderHeaders)
            .then((response) => {
                const appointmentRequests = response.data;
                const totalPages = Math.ceil(appointmentRequests.length / TOTAL_PER_PAGE);

                if (appointmentRequests.length === 0) {
                    this.setState({
                        appointmentRequests: appointmentRequests,
                    });
                } else
                    appointmentRequests.map((appointmentRequest, index) => {
                        usersStorage.getUserByUserID(appointmentRequest.AppointmentDetail.clientId, this.serviceProviderHeaders)
                            .then(user => {
                                appointmentRequest.clientName = user.fullname;
                                appointmentRequest.optionalTimes = JSON.parse(appointmentRequest.optionalTimes);

                                let appointmentRequestEvent = {
                                    id: appointmentRequest.requestId,
                                    title: appointmentRequest.clientName,
                                    allDay: false,
                                    start: null,
                                    end: null,
                                    appointmentRequest: appointmentRequest,
                                    // color:'#b7d2ff',
                                    backgroundColor: '#45b0d9',
                                };

                                let appointmentRequestsEvents = this.state.appointmentRequests;
                                if (appointmentRequestsEvents.filter(item => item.id === appointmentRequest.requestId).length === 0) {
                                    appointmentRequestsEvents.push(appointmentRequestEvent);
                                    this.setState({
                                        appointmentRequests: appointmentRequestsEvents
                                    });
                                }

                            });
                    });

                this.setState({
                    page: 0,
                    totalPages,
                });
            });
    }


    componentWillReceiveProps({location = {}}) {
        if (location.pathname === '/appointments' && location.pathname !== this.props.location.pathname) {
            this.setState({
                appointments: [],
                appointmentRequests: [],
            });
            this.getServiceProviderAppointments();
            this.getServiceProviderAppointmentRequests();
        }
    }

    getServiceProviderAppointments() {
        appointmentsStorage.getServiceProviderAppointments(this.serviceProviderId, this.serviceProviderHeaders)
            .then((response) => {
                this.setState({appointments: []});

                const appointments = response.data;
                const totalPages = Math.ceil(appointments.length / TOTAL_PER_PAGE);

                if (appointments.length === 0) {
                    this.setState({
                        appointments: appointments,
                    });
                } else
                    appointments.map((appointment, index) => {
                        usersStorage.getUserByUserID(appointment.AppointmentDetail.clientId, this.serviceProviderHeaders)
                            .then(user => {
                                let appointmentEvent = {};
                                appointment.clientName = user.fullname;

                                appointmentEvent.id = appointment.appointmentId;
                                appointmentEvent.title = user.fullname;
                                appointmentEvent.allDay = false;
                                appointmentEvent.start = moment(appointment.startDateAndTime);
                                appointmentEvent.end = moment(appointment.endDateAndTime);
                                appointmentEvent.appointment = appointment;

                                let appointmentsEvents = this.state.appointments;
                                appointmentsEvents.push(appointmentEvent);
                                this.setState({
                                    appointments: appointmentsEvents
                                });
                            });
                    });

                this.setState({
                    page: 0,
                    totalPages,
                });

            });
    }

    setPage(page) {
        return () => {
            this.setState({page});
        };
    }

    decrementPage() {
        const {page} = this.state;

        this.setState({page: page - 1});
    }

    incrementPage() {
        const {page} = this.state;

        this.setState({page: page + 1});
    }

    handleDelete(appointmentId) {
        const {appointments} = this.state;

        this.setState({
            appointments: appointments.filter(a => a.id !== appointmentId),
        });
    }

    hoverOnAppointmentRequest = (appointmentRequest) => {
        let isAppointmentsWithOptional = this.state.appointments.filter(obj => obj.id === appointmentRequest.requestId && obj.status === 'optional');
        if (isAppointmentsWithOptional.length > 0)
            return;

        this.setState({appointmentRequestHoovering: appointmentRequest});
        let shadowAppointments = [];
        if (Array.isArray(appointmentRequest.optionalTimes)) {
            appointmentRequest.optionalTimes.map(datesTimes => {
                datesTimes.hours.map(time => {
                    shadowAppointments.push(
                        {
                            id: appointmentRequest.requestId,
                            title: appointmentRequest.clientName,
                            allDay: false,
                            start: moment(moment(datesTimes.date).format("YYYY-MM-DD") + ' ' + time.startHour).toDate(),
                            end: moment(moment(datesTimes.date).format("YYYY-MM-DD") + ' ' + time.endHour).toDate(),
                            appointmentRequest: appointmentRequest,
                            status: "optional",
                            // color:'#b7d2ff',
                            backgroundColor: '#45b0d9',
                        }
                    )
                })
            });
            shadowAppointments.push.apply(shadowAppointments, this.state.appointments);
            if (Array.isArray(shadowAppointments)) {
                this.setState({appointments: shadowAppointments});
            }
        }
    };

    hoverOffAppointmentRequest = (appointmentRequest) => {
        let appointmentsWithoutOptional = this.state.appointments.filter(obj => obj.id !== appointmentRequest.requestId || obj.status !== 'optional');
        this.setState({appointments: appointmentsWithoutOptional});
    };

    onSelectEvent = event => {
        console.log('onSelectEvent=event  ', event);
        // if (event.appointmentId === this.state.highlightTableRow)
        //     this.setState({highlightTableRow: null});
        // else {
        //     this.setState({highlightTableRow: event.appointmentId});
        //
        // }
        /*this.props.history.push(`${this.props.match.path}/${event.appointment.appointmentId}`, {
            appointment: event.appointment
        });*/
    };

    onSelectSlot = (start, end) => {
        let slotInfo = {start: start, end: end};

        /*this.props.history.push(`${this.props.match.path}/set`, {
            slotInfo: slotInfo,
        });*/
    };

    approveAppointmentRequest = (appointmentRequestEventDropped) => {
        var appointmentRequests = this.state.appointmentRequests;
        appointmentsStorage.approveAppointmentRequestById(appointmentRequestEventDropped.appointmentRequest, this.serviceProviderHeaders)
            .then(response => {
                console.log('appointmentRequest approved ', response);

                let updatedAppointmentsRequests = appointmentRequests.filter(function (obj) {
                    return obj.appointmentRequest.requestId !== appointmentRequestEventDropped.appointmentRequest.requestId;
                });
                this.setState({appointmentRequests: updatedAppointmentsRequests})
            });
    };

    onDropAppointmentRequest = (appointmentRequestEvent) => {
        let appointmentRequestDropped = {
            id: appointmentRequestEvent.id,
            title: appointmentRequestEvent.title,
            start: appointmentRequestEvent.start,
            end: moment(appointmentRequestEvent.start).add(2, 'h'),
            allDay: appointmentRequestEvent.allDay,
            appointmentRequest: appointmentRequestEvent.appointmentRequest,
        };
        /*this.props.history.push(`${this.props.match.path}/set`, {
            appointmentRequestDropped: appointmentRequestDropped,
        });*/
    };

    updateAfterMoveOrResizeEvent(event) {
        let events = this.state.appointments;

        let appointment = event.appointment;
        appointment.startDateAndTime = moment(event.start).format();
        appointment.endDateAndTime = moment(event.end).format();

        const idx = events.indexOf(event);

        const updatedEvent = {...event, appointment: appointment};

        const nextEvents = [...events];
        nextEvents.splice(idx, 1, updatedEvent);

        this.updateAppointment(appointment, nextEvents);
    }

    updateAppointment = (appointment, nextEvents) => {
        let event = {
            appointmentId: appointment.appointmentId,
            date: moment(appointment.startDateAndTime).format('YYYY-MM-DD'),
            startTime: moment(appointment.startDateAndTime).format('HH:mm'),
            endTime: moment(appointment.endDateAndTime).format('HH:mm'),
            remarks: appointment.remarks,
            subject: appointment.subject,
            clientId: appointment.clientId,
        };

        appointmentsStorage.updateAppointment(event, this.serviceProviderHeaders)
            .then((response) => {
                this.setState({
                    appointments: nextEvents,
                })
            })
    };


    render() {
        const {appointments, page, totalPages} = this.state;
        const startIndex = page * TOTAL_PER_PAGE;

        console.log("AppointmentsReportPage appointments ", appointments);

        return (
            <div>
                <div>
                    <Helmet>
                        <title>Meshekle | Appointments Report</title>
                    </Helmet>
                    <Grid stretched padded>
                        <Grid.Row>
                            <Header as="h1"
                                    floated="right">{strings.mainPageStrings.REPORT_PAGE_TITLE}</Header>
                        </Grid.Row>

                        <Link to={{
                            pathname: `${this.props.match.url}/serviceProvider/settings`,
                            state: {serviceProviderId: this.serviceProviderId, users: []}
                        }}>
                            <Button positive icon>
                                <Icon name="settings"/>
                                &nbsp;&nbsp;
                                {strings.mainPageStrings.SETTINGS_PAGE_TITLE}
                            </Button>
                        </Link>
                        <Link to='/appointments'>
                            <Button positive icon>
                                <Icon name="calendar"/>
                                &nbsp;&nbsp;
                                {strings.mainPageStrings.BACK_TO_APPOINTMENTS_PAGE_TITLE}
                            </Button>
                        </Link>

                        <Grid.Row columns='equal'>
                            <Table celled striped textAlign='right' selectable sortable>
                                <Table.Header>
                                    <Table.Row>
                                        <Table.HeaderCell></Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.CLIENT_ID}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.SERVICE_PROVIDER_ID}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.ROLE}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.SUBJECT}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.STATUS}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.DATE}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.START_TIME}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.END_TIME}</Table.HeaderCell>
                                        <Table.HeaderCell>{strings.appointmentsPageStrings.REMARKS}</Table.HeaderCell>
                                    </Table.Row>
                                </Table.Header>
                                <Table.Body>
                                    {appointments.slice(startIndex, startIndex + TOTAL_PER_PAGE).map(appointmentEvent =>
                                        (<Table.Row key={appointmentEvent.appointment.appointmentId}
                                                    positive={this.state.highlightTableRow === appointmentEvent.appointment.appointmentId}>
                                            <Table.Cell>
                                                <Icon link name='edit'
                                                      onClick={() => this.props.history.push(`${this.props.match.path}/${appointmentEvent.appointment.appointmentId}/edit`, {
                                                          appointment: appointmentEvent.appointment,
                                                          openedFrom: "AppointmentsReportPage"
                                                      })}
                                                />
                                            </Table.Cell>
                                            <Table.Cell>{appointmentEvent.appointment.AppointmentDetail.clientId}</Table.Cell>
                                            <Table.Cell>{appointmentEvent.appointment.AppointmentDetail.serviceProviderId}</Table.Cell>
                                            <Table.Cell>{appointmentEvent.appointment.AppointmentDetail.role}</Table.Cell>
                                            <Table.Cell>{JSON.parse(appointmentEvent.appointment.AppointmentDetail.subject).join(", ")}</Table.Cell>
                                            <Table.Cell>{appointmentEvent.appointment.status}</Table.Cell>
                                            <Table.Cell>{new Date(appointmentEvent.appointment.startDateAndTime).toISOString().split('T')[0]}</Table.Cell>
                                            <Table.Cell>{new Date(appointmentEvent.appointment.startDateAndTime).toISOString().split('T')[1].split('.')[0].slice(0, -3)}</Table.Cell>
                                            <Table.Cell>{new Date(appointmentEvent.appointment.endDateAndTime).toISOString().split('T')[1].split('.')[0].slice(0, -3)}</Table.Cell>
                                            <Table.Cell>{appointmentEvent.appointment.remarks}</Table.Cell>
                                        </Table.Row>),
                                    )}
                                </Table.Body>
                                <Table.Footer>
                                    <Table.Row>
                                        <Table.HeaderCell colSpan={10}>
                                            <Menu floated="left" pagination>
                                                {page !== 0 && <Menu.Item as="a" icon onClick={this.decrementPage}>
                                                    <Icon name="right chevron"/>
                                                </Menu.Item>}
                                                {times(totalPages, n =>
                                                    (<Menu.Item as="a" key={n} active={n === page}
                                                                onClick={this.setPage(n)}>
                                                        {n + 1}
                                                    </Menu.Item>),
                                                )}
                                                {page !== (totalPages - 1) &&
                                                <Menu.Item as="a" icon onClick={this.incrementPage}>
                                                    <Icon name="left chevron"/>
                                                </Menu.Item>}
                                            </Menu>
                                        </Table.HeaderCell>
                                    </Table.Row>
                                </Table.Footer>
                            </Table>

                        </Grid.Row>
                    </Grid>
                </div>


                <div>
                    <Switch>
                        <Route exec path={`${this.props.match.url}/serviceProvider/settings`}
                               component={ServiceProviderEdit}/>
                        <Route exec path={`${this.props.match.path}/requests/:appointmentRequestId`}
                               component={AppointmentRequestInfo}/>
                        <Route exec path={`${this.props.match.path}/set`} render={(props) => (
                            <AppointmentAdd {...props}
                                            approveAppointmentRequest={(appointmentRequestEvent) => this.approveAppointmentRequest(appointmentRequestEvent)}/>
                        )}/>
                        <Route exec path={`${this.props.match.path}/:appointmentId`}
                               component={AppointmentInfo}/>
                        <Route exec path={`${this.props.match.path}/:appointmentId/edit`}
                               component={AppointmentEdit}/>

                        {/*{*/}
                        {/*(this.props.location.pathname === window.location.pathname) ?*/}
                        {/*<Redirect to={`${this.props.match.path}`}/>*/}
                        {/*: null}*/}
                        {/*}*/}
                    </Switch>
                </div>
            </div>
            // </Grid>*/
        );
    }
}

export default AppointmentsReportPage

