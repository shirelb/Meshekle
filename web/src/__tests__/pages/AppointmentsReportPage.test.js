import React from 'react';
import {shallow} from "enzyme/build";
import {setupComponent} from "../testHelpers";
import PageNotFound from "../../pages/pageNotFound404/PageNotFound";
import AppointmentsReportPage from '../../pages/appointmentsManagementPage/AppointmentsReportPage';
import store from 'store';
import {createMemoryHistory} from 'history';
import serviceProvidersStorage from "../../storage/serviceProvidersStorage";
import appointmentsStorage from "../../storage/appointmentsStorage";
import usersStorage from "../../storage/usersStorage";
import users from "../jsons/users";
import serviceProviders from "../jsons/serviceProviders";
import appointmentsOf549963652 from "../jsons/appointmentsServiceProvider549963652";
import appointmentRequestsOf549963652 from "../jsons/appointmentRequestsServiceProvider549963652";
import strings from "../../shared/strings";
import helpers from "../../shared/helpers";

jest.mock("../../shared/helpers");
jest.mock("store");
jest.mock("../../storage/usersStorage");
jest.mock("../../storage/serviceProvidersStorage");
jest.mock("../../storage/appointmentsStorage");


describe("AppointmentsReportPage should", () => {
    let wrapper = null;
    let componentInstance = null;
    const history = createMemoryHistory({
        initialEntries: ['/', '/home', '/appointments', '/appointments/report'],
        initialIndex: 3,
        keyLength: 10,
        getUserConfirmation: null
    });
    const props = {
        history: history,
        location: {
            pathname: '/appointments/report'
        },
        match: {
            isExact: true,
            path: '/appointments/report',
            url: '/appointments/report',
        }
    };
    const mockStore = {
        serviceProviderId: "549963652",
        userId: "549963652",
    };

    store.get.mockImplementation((key) => mockStore[key]);
    usersStorage.getUsers.mockResolvedValue(users);
    usersStorage.getUserByUserID.mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviderUserDetails.mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById.mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    appointmentsStorage.getServiceProviderAppointments.mockResolvedValue({data: appointmentsOf549963652});
    appointmentsStorage.getServiceProviderAppointmentRequests.mockImplementation(() => Promise.resolve({data: JSON.parse(JSON.stringify(appointmentRequestsOf549963652))}));

    beforeAll(async (done) => {
        done();
    });

    afterAll(() => {
    });

    beforeEach(async (done) => {
        await jest.clearAllMocks();
        done();
    });

    afterEach(() => {
    });

    test('match snapshot', async () => {
        const props = {
            location: {
                pathname: '/appointments/report'
            },
            match: {
                isExact: true,
                path: '/appointments/report',
                url: '/appointments/report',
            }
        };

        const arrResponse = setupComponent('shallow', AppointmentsReportPage, null, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance).toMatchSnapshot();
    });

    test("renders AppointmentsReportPage for /appointments/report", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AppointmentsReportPage)).toHaveLength(1);
        expect(wrapper.find(PageNotFound)).toHaveLength(0);
    });

    test("mounted with the right data", async () => {
        expect(componentInstance.state.page).toEqual(0);
        expect(componentInstance.state.totalPages).toEqual(1);
        expect(componentInstance.state.appointments.length).toEqual(3);
    });

    test("render with what the user see", async () => {
        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const mountWrapper = wrapper.find('AppointmentsReportPage');

        expect(mountWrapper.find('Grid')).toHaveLength(1);
        expect(mountWrapper.find('Header')).toHaveLength(1);
        expect(mountWrapper.find('Header').get(0).props.children).toEqual(strings.mainPageStrings.REPORT_PAGE_TITLE);
        expect(mountWrapper.find('Link')).toHaveLength(1);
        expect(mountWrapper.find('Link').first().find('Button')).toHaveLength(1);
        expect(mountWrapper.find('Link').first().find('Button').props().children[2]).toEqual(strings.mainPageStrings.SETTINGS_PAGE_TITLE);
        expect(mountWrapper.find('Button')).toHaveLength(4);
        expect(mountWrapper.find('Button').at(1).props().children[2]).toEqual(strings.mainPageStrings.BACK_TO_APPOINTMENTS_PAGE_TITLE);
        expect(mountWrapper.find('Button').at(2).props().children[1]).toEqual('   יצא לPDF');
        expect(mountWrapper.find('Table')).toHaveLength(1);
        expect(mountWrapper.find('TableRow')).toHaveLength(6);
        expect(mountWrapper.find('TableHeader')).toHaveLength(2);
        expect(mountWrapper.find('TableRow').at(0).props().children).toHaveLength(10);
        expect(mountWrapper.find('TableRow').at(0).props().children[1].props.children).toEqual(strings.appointmentsPageStrings.CLIENT_ID);
        expect(mountWrapper.find('TableRow').at(0).props().children[2].props.children).toEqual(strings.appointmentsPageStrings.CLIENT_NAME);
        expect(mountWrapper.find('TableRow').at(0).props().children[3].props.children).toEqual(strings.appointmentsPageStrings.ROLE);
        expect(mountWrapper.find('TableRow').at(0).props().children[4].props.children).toEqual(strings.appointmentsPageStrings.SUBJECT);
        expect(mountWrapper.find('TableRow').at(0).props().children[5].props.children).toEqual(strings.appointmentsPageStrings.STATUS);
        expect(mountWrapper.find('TableRow').at(0).props().children[6].props.children).toEqual(strings.appointmentsPageStrings.DATE);
        expect(mountWrapper.find('TableRow').at(0).props().children[7].props.children).toEqual(strings.appointmentsPageStrings.START_TIME);
        expect(mountWrapper.find('TableRow').at(0).props().children[8].props.children).toEqual(strings.appointmentsPageStrings.END_TIME);
        expect(mountWrapper.find('TableRow').at(0).props().children[9].props.children).toEqual(strings.appointmentsPageStrings.REMARKS);
        expect(mountWrapper.find('TableRow').at(1).props().children).toHaveLength(10);
        expect(mountWrapper.find('TableRow').at(1).props().children[1].props.children).toHaveLength(2);
        expect(mountWrapper.find('TableRow').at(1).props().children[1].props.children[0].props.name).toEqual('filter');
        expect(mountWrapper.find('TableRow').at(1).props().children[1].props.children[1].props.className).toEqual('filterInput');
        expect(mountWrapper.find('TableBody')).toHaveLength(1);
        expect(mountWrapper.find('TableBody').props().children).toHaveLength(3);
        expect(mountWrapper.find('TableFooter')).toHaveLength(1);
        expect(mountWrapper.find('TableFooter').find('MenuItem')).toHaveLength(1);
    });

    test("open ServiceProviderEdit on click Settings", async () => {
        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const settingsBtn = wrapper.find('AppointmentsReportPage').find('Link').first();
        await settingsBtn.simulate('click', {button: 0});

        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/appointments/report/serviceProvider/settings')
        expect(wrapper.find('ServiceProviderEdit')).toHaveLength(1);
    });

    test.skip("open ReportPage on click Report", async () => {
        const arrResponse = await setupComponent('shallow', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const reportBtn = wrapper.find('AppointmentsReportPage').dive().find('Button').at(1);
        await reportBtn.simulate('click', {button: 0});

        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/appointments');
        expect(wrapper.find('AppointmentsManagementPage')).toHaveLength(1);
    });

    test("save as PDF on click Report", async () => {
        helpers.exportToPDF = jest.fn();

        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const PDFBtn = wrapper.find('AppointmentsReportPage').find('Button').at(2);
        await PDFBtn.simulate('click', {button: 0});

        expect(helpers.exportToPDF).toHaveBeenCalledWith('MeshekleAppointmentsReport', 'divToPrint', 'landscape')
    });

    test("render page of appointments on click of another page", async () => {
        const setPageSpy = jest.spyOn(AppointmentsReportPage.prototype, 'setPage');

        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const page1Btn = wrapper.find('AppointmentsReportPage').find('MenuItem');
        await page1Btn.simulate('click', {button: 0});

        expect(setPageSpy).toHaveBeenCalled();
    });

    test.skip("render next page of appointments on click of another page", async () => {
        const incrementPageSpy = jest.spyOn(AppointmentsReportPage.prototype, 'incrementPage');

        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const nextPageBtn = wrapper.find('AppointmentsReportPage').find('MenuItem').at(2);
        await nextPageBtn.simulate('click', {button: 0});

        expect(incrementPageSpy).toHaveBeenCalled();
    });

    test.skip("render previous page of appointments on click of another page", async () => {
        const decrementPageSpy = jest.spyOn(AppointmentsReportPage.prototype, 'decrementPage');

        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const previousPageBtn = wrapper.find('AppointmentsReportPage').find('MenuItem').at(2);
        await previousPageBtn.simulate('click', {button: 0});

        expect(decrementPageSpy).toHaveBeenCalled();
    });

    test("render /appointments/report/428/edit on edit appointment", async () => {
        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const editBtn = wrapper.find('TableBody').find('Icon').at(0);
        await editBtn.simulate('click', {button: 0});

        expect(componentInstance.props.history.location.pathname).toEqual('/appointments/report/428/edit');
    });

    test("sort the table by userId when userId column clicked", async () => {
        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const handleSortSpy = jest.spyOn(componentInstance, 'handleSort');

        const sortBtn = wrapper.find('TableHeaderCell').at(1);
        await sortBtn.simulate('click', {button: 0});

        expect(handleSortSpy).toHaveBeenCalled();
        expect(componentInstance.state.appointments[0].clientId).toEqual("561933550");
        expect(componentInstance.state.appointments[1].clientId).toEqual("580624494");
        expect(componentInstance.state.appointments[2].clientId).toEqual("616314141");
        expect(componentInstance.state.column).toEqual("clientId");
        expect(componentInstance.state.direction).toEqual("ascending");
    });

    test("filter the table by userId when userId column clicked", async () => {
        const arrResponse = await setupComponent('mount', AppointmentsReportPage, history, props, "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const handleFilterSpy = jest.spyOn(componentInstance, 'handleFilter');

        wrapper.find('TableHeaderCell').at(11).find('Input').props().onChange({
            target: {
                value: "580624494"
            }
        });

        expect(handleFilterSpy).toHaveBeenCalled();
        expect(componentInstance.state.appointments).toHaveLength(1);
        expect(componentInstance.state.appointments[0].clientId).toEqual("580624494");
        expect(componentInstance.state.filterColumnsAndTexts.clientId).toEqual("580624494");
        expect(componentInstance.state.filterColumnsAndTexts.clientName).toEqual("");
        expect(componentInstance.state.filterColumnsAndTexts.serviceProviderId).toEqual("");
        expect(componentInstance.state.filterColumnsAndTexts.role).toEqual("");
        expect(componentInstance.state.filterColumnsAndTexts.subject).toEqual("");
        expect(componentInstance.state.filterColumnsAndTexts.date).toEqual("");
    });

});
