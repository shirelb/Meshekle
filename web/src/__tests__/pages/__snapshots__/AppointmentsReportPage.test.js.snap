// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentsReportPage should match snapshot 1`] = `undefined`;
